src/NosCore.Database/DAL/ItemInstanceDAO.cs

```diff
@@ -222,52 +222,42 @@ public SaveResult InsertOrUpdate(IEnumerable<IItemInstanceDto> dtos)
             try
             {
                 var context = DataAccessHelper.Instance.CreateContext();
+
                 context.ChangeTracker.AutoDetectChangesEnabled = false;
 
                 var dbset = context.Set<ItemInstance>();
                 var entitytoadd = new List<ItemInstance>();
+                var list = new List<Tuple<ItemInstance, object>>();
+
                 foreach (var dto in dtos)
                 {
-                    var entity = dto.GetType().Name == "BoxInstance" ? dto.Adapt<BoxInstance>()
+                    list.Add(new Tuple<ItemInstance, object>(dto.GetType().Name == "BoxInstance" ? dto.Adapt<BoxInstance>()
                         : dto.GetType().Name == "SpecialistInstance" ? dto.Adapt<SpecialistInstance>()
                         : dto.GetType().Name == "WearableInstance" ? dto.Adapt<WearableInstance>()
                         : dto.GetType().Name == "UsableInstance" ? dto.Adapt<UsableInstance>()
-                        : dto.Adapt<ItemInstance>();
-
-                    var value = _primaryKey.GetValue(dto, null);
-
-                    ItemInstance entityfound = null;
-                    if (value is object[] objects)
-                    {
-                        entityfound = dbset.Find(objects);
-                    }
-                    else
-                    {
-                        entityfound = dbset.Find(value);
-                    }
-
-                    entity = entity is BoxInstance ? entity.Adapt<BoxInstanceDto>().Adapt<BoxInstance>() :
-                        entity is SpecialistInstance
-                            ? entity.Adapt<SpecialistInstanceDto>().Adapt<SpecialistInstance>() :
-                            entity is WearableInstance
-                                ? entity.Adapt<WearableInstanceDto>().Adapt<WearableInstance>() :
-                                entity is UsableInstance ? entity.Adapt<UsableInstanceDto>().Adapt<UsableInstance>()
-                                    :
-                                    entity.Adapt<ItemInstanceDto>().Adapt<ItemInstance>();
+                        : dto.Adapt<ItemInstance>(), _primaryKey.GetValue(dto, null)));
+                }
+                var ids = list.Select(s => s.Item2).ToArray();
+                var dbkey = typeof(ItemInstance).GetProperty(_primaryKey.Name);
+                var entityfounds = dbset.FindAllAsync(dbkey, ids).ToList();
 
+                foreach (var dto in list)
+                {
+                    var entity = dto.Item1 is BoxInstance ? dto.Item1.Adapt<BoxInstanceDto>().Adapt<BoxInstance>() : dto.Item1 is SpecialistInstance
```

Comment:
Not very important here, but such a ternary is not really readable, I would rather use proper if's even if statements or a switch instead of a 5 lines ternary condition

Reply:
just move the old one :) this whole file is not really a good thing but required